### PR TITLE
Updated code (contrastive learning) 

### PR DESCRIPTION
**Pull request by alishbaimran**
I made the following changes:
—> Updated triplet code such that during prediction evaluation can be done for specific pre-defined cells. 
—> Changed stride to be the same as stem kernel size which was also changed to (5, 4, 4).
—> Output file name is configurable for the prediction. 
—> Updated training_script.py and predict.py

To-do:
—> Should be merged before any other branches merged into main.

### CODE REVIEW

File: viscy/light/engine.py [hunk 1]
```diff
@@ -745,6 +751,15 @@ def predict_step(self, batch: TripletSample, batch_idx, dataloader_idx=0):
...
     def on_predict_epoch_end(self) -> None:
+        if (
+            not self.features_output_path
+            or not self.projections_output_path
+            or not self.metadata_output_path
+        ):
+            raise ValueError(
+                "Output paths for features, projections, and metadata must be provided."
+            )
+
```
Comment (ziw-liu) [inline viscy/light/engine.py:762]: Should be in `on_predict_start` instead?

Comment (alishbaimran) [reply viscy/light/engine.py]: fixed

File: viscy/data/triplet.py [hunk 2]
```diff
@@ -56,6 +56,9 @@ def __init__(
         positive_transform: DictTransform | None = None,
         negative_transform: DictTransform | None = None,
         fit: bool = True,
+        predict_cells: bool = False,
+        include_fov_names: list[str] = None,
+        include_track_ids: list[int] = None,
```
Comment (ziw-liu) [inline viscy/data/triplet.py:61]:
```suggestion
        include_fov_names: list[str] | None = None,
        include_track_ids: list[int] | None = None,
```

File: viscy/data/triplet.py [hunk 3]
```diff
@@ -182,6 +202,9 @@ def __init__(
         normalizations: list[MapTransform] = [],
         augmentations: list[MapTransform] = [],
         caching: bool = False,
+        predict_cells: bool = False,
+        include_fov_names: list[str] = None,
+        include_track_ids: list[int] = None,
```
Comment (ziw-liu) [inline viscy/data/triplet.py:207]:
```suggestion
        include_fov_names: list[str] | None = None,
        include_track_ids: list[int] | None = None,
```

File: viscy/light/engine.py [hunk 4]
```diff
@@ -734,9 +740,19 @@ def configure_optimizers(self):
...
+            and self.metadata_output_path
+        ):
+            raise ValueError(
+                "Output paths for features, projections, and metadata must be provided."
+            )
+
     def predict_step(self, batch: TripletSample, batch_idx, dataloader_idx=0):
-        print("running predict step!")
         """Prediction step for extracting embeddings."""
+
```
Comment (ziw-liu) [inline viscy/light/engine.py:755]:
```suggestion
```

File: viscy/data/triplet.py [hunk 5]
```diff
@@ -68,7 +71,13 @@ def __init__(
...
         self.yx_patch_size = initial_yx_patch_size
+        self.predict_cells = predict_cells
+        self.include_fov_names = include_fov_names or []
+        self.include_track_ids = include_track_ids or []
         self.tracks = self._filter_tracks(tracks_tables)
+        self.tracks = (
+            self._specific_cells(self.tracks) if self.predict_cells else self.tracks
+        )
 
     def _filter_tracks(self, tracks_tables: list[pd.DataFrame]) -> pd.DataFrame:
```
Comment (mattersoflight) [inline viscy/data/triplet.py:82]: add a short docstring describing the purpose.

File: applications/contrastive_phenotyping/predict.py [hunk 6]
```diff
@@ -17,62 +17,103 @@
...
+        b_max=1.0,
+        clip=False,  
+        relative=False, 
+        channel_wise=False,  
+        dtype=None, 
+        allow_missing_keys=False  
+    ),
 ]
 
 def main(hparams):
```
Comment (mattersoflight) [inline applications/contrastive_phenotyping/predict.py:47]: Let's clean up the predict.py in the next PR

File: viscy/light/engine.py [hunk 7]
```diff
@@ -581,10 +581,13 @@ def __init__(
...
         example_input_yx_shape: Sequence[int] = (256, 256),
         in_stack_depth: int = 15,
-        stem_kernel_size: tuple[int, int, int] = (5, 3, 3),
+        stem_kernel_size: tuple[int, int, int] = (5, 4, 4),
         embedding_len: int = 256,
         predict: bool = False,
         tracks_path: str = "data/tracks",
+        features_output_path: str = "",
+        projections_output_path: str = "",
+        metadata_output_path: str = "",
```
Comment (mattersoflight) [inline viscy/light/engine.py:590]: Looks like these three paths are needed to save predictions. I suggest renaming these three to `predict_output_path` and assign these three variables a consistent names, predict_output_path + `features.npy`, `projecttions.npy`, `meta.csv` ...

File: applications/contrastive_phenotyping/training_script.py [hunk 8]
```diff
@@ -233,30 +187,21 @@ def main(hparams):
...
-    # Aggregates the losses and performs optimization steps
     trainer.fit(model, datamodule=data_module)
 
     # # Validate the model
     trainer.validate(model, datamodule=data_module)
 
-    # # Test the model
-    # trainer.test(model, datamodule=data_module)
-
 # Argument parser for command-line options
```
Comment (mattersoflight) [inline applications/contrastive_phenotyping/training_script.py:198]: @alishbaimran In the next PR, this training script should be turned into a config for use with lightning CLI and then deleted.

**Review by mattersoflight — COMMENTED** [review]
I am merging this PR to unblock the rest of the improvements in model construction. Let's revisit comments in this PR to:
* delete applications/contrastive_phenotyping/training_script.py
* clean up predict.py.
* document key methods in viscy.data.triplet.py
* clean up excess input parameters to viscy.light.engine.ContrastiveModule

**Review by mattersoflight — APPROVED** [review]

**Review by mattersoflight — APPROVED** [review]

**Pull request merged**